solr/core/src/java/org/apache/solr/search/facet/FacetProcessor.java

```diff
@@ -190,16 +191,23 @@ private void handleDomainChanges() throws IOException {
   }
 
   private void handleFilterExclusions() throws IOException {
+    List<Query> qlist = getContextQueries();
+
+    // recompute the base domain
+    fcontext.base = fcontext.searcher.getDocSet(qlist);
+  }
+
+  protected List<Query> getContextQueries() {
     List<String> excludeTags = freq.domain.excludeTags;
 
     if (excludeTags == null || excludeTags.size() == 0) {
-      return;
+      return Collections.emptyList();
     }
 
     Set<Query> excludeSet = QueryUtils.getTaggedQueries(fcontext.req, excludeTags);
 
     if (excludeSet.isEmpty()) {
-      return;
+      return Collections.emptyList();
```

Comment:
We can't shortcircuit here (this will probably fail existing solr tests, I think?). The reason stock solr can shortcircuit is that if there are no tag/exclusions, this facet just inherits parent `fcontext`'s `fcontext.base` DocSet. But with this PR in Solr we're recomputing `fcontext.base` every time (which we can/should avoid). It's also returning empty query list, which will end up calculating it wrong I think anyway (and empty query list is not what we want form the FunnelFacetProcessor side either).

It's a little tricky; we want to:
1. avoid recomputing `fcontext.base` if it's not necessary
2. ideally avoid building query list if the caller doesn't care about it (solr-internal for the the "no exclusion" case, and
3. always get query list (respecting exclusions) if the caller _does_ care about something other than computing `fcontext.base` (i.e., when called from FunnelFacetProcessor)

Reply:
That makes sense, I think I've resolved those issues now.

Reply:
Is there a way that we can avoid calling `QueryUtils.getTaggedQueries(...)` twice for the `handleFilterExclusions()` path?

Reply:
Just changed that to pass the excludeSet into getContextQueries(). If it is null, it will recompute so from FunnelFacetProcessor for example, we can just call getContextQueries(null).